### PR TITLE
[prim_sha2,rtl] Add prim_sha2 RTL and instantiation in DMA and HMAC

### PR DESCRIPTION
**Pull request by gdessouky**
This PR has 5 commits:

1. `[prim_sha2,rtl] Baseline RTL implementation`  which is already also included in the reduced PR #20675 
2. `[prim_sha2,rtl] Add RTL implementation + update core + lint` which is already also included in the reduced PR #20675 
3. `[hmac/rtl] Update HMAC to instantiate SHA-2 prim` which instantiates the 256-bit variant of the SHA-2 prim in the HMAC after reaching the same coverage numbers and pass rates.
4. `[hmac,dv] Update HMAC UNR-generated coverage exclusions` which updates the coverage exclusions list for HMAC to update the coverage evaluation numbers.
5. ` [dma,rtl] Update DMA to instantiate SHA-2` which instantiates the multi-mode variant of the SHA-2 prim in DMA after reaching same coverage numbers and pass rates.

With these changes, this PR checks out items 1-3 in this TODO list https://github.com/lowRISC/opentitan-integrated/issues/451

Note that PR #20675 is a reduced PR spun-off from this one that isolates the SHA-2 prim RTL changes, i.e., first 2 commits of this PR.

We will need to decide if this will be cherry-picked into discrete for PROD @moidx.

### CODE REVIEW

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (a-will) [inline hw/ip/prim/rtl/prim_sha2.sv:442]: There's no assignment to `sha_st_d` here, leading to a latch construction.

Maybe add

```systemverilog
sha_st_d = sha_st_q;
```

as a default alongside the others above?

**Comment by vogelpi** [comment]
Hey @gdessouky ,
thanks for the PR. I haven't reviewed in depth yet. In principle, I don't mind the changes but I think we need to be careful regarding changing the HMAC hardware block. Because this block is at D3 already (so no RTL changes at this point other than bug fixes) and we need it for the production tapeout for Earl Grey. I fear if we merge this PR as is, it will be very hard to merge/cherry-pick `integrated_dev` into master before the next tapeout.

I think also @msfschaffner and @andreaskurth should weigh in on this (please wait for their feedback before making big changes to the PR). I see two options:
- Either we should wait with updating HMAC at the moment. This would still allow merging the `prim_sha2` and the DMA changes of this PR.
- We merge also the HMAC IP changes but then you should maybe run a full coverage regression with DVSIM and ensure coverage as well as pass/fail rates stay the same.

I am in favor of the first option but others should weigh in as well. Related to this, we should discuss this in the Silicon working group, e.g., tomorrow evening. Can you please liaise with the Silicon WG chairs @GregAC and @msfschaffner to make sure it gets on the agenda?

**Comment by a-will** [comment]
> Hey @gdessouky , thanks for the PR. I haven't reviewed in depth yet. In principle, I don't mind the changes but I think we need to be careful regarding changing the HMAC hardware block. Because this block is at D3 already (so no RTL changes at this point other than bug fixes) and we need it for the production tapeout for Earl Grey. I fear if we merge this PR as is, it will be very hard to merge/cherry-pick `integrated_dev` into master before the next tapeout.
> 
> I think also @msfschaffner and @andreaskurth should weigh in on this (please wait for their feedback before making big changes to the PR). I see two options:
> 
> * Either we should wait with updating HMAC at the moment. This would still allow merging the `prim_sha2` and the DMA changes of this PR.
> * We merge also the HMAC IP changes but then you should maybe run a full coverage regression with DVSIM and ensure coverage as well as pass/fail rates stay the same.
> 
> I am in favor of the first option but others should weigh in as well. Related to this, we should discuss this in the Silicon working group, e.g., tomorrow evening. Can you please liaise with the Silicon WG chairs @GregAC and @msfschaffner to make sure it gets on the agenda?

I feel like this is a significant enough change to warrant a new version of the HMAC IP (a v2), regardless of when this PR merges. Since it's a new version anyway, that ought to allow merging it into `integrated_dev` whenever. This PR should include that version change, though.

When we start bringing Darjeeling into the fold on master, we'll realistically be supporting both versions for quite awhile, unless we decide to upgrade earlgrey's for final silicon.

**Comment by gdessouky** [comment]
> Hey @gdessouky , thanks for the PR. I haven't reviewed in depth yet. In principle, I don't mind the changes but I think we need to be careful regarding changing the HMAC hardware block. Because this block is at D3 already (so no RTL changes at this point other than bug fixes) and we need it for the production tapeout for Earl Grey. I fear if we merge this PR as is, it will be very hard to merge/cherry-pick `integrated_dev` into master before the next tapeout.
> 
> I think also @msfschaffner and @andreaskurth should weigh in on this (please wait for their feedback before making big changes to the PR). I see two options:
> 
> * Either we should wait with updating HMAC at the moment. This would still allow merging the `prim_sha2` and the DMA changes of this PR.
> * We merge also the HMAC IP changes but then you should maybe run a full coverage regression with DVSIM and ensure coverage as well as pass/fail rates stay the same.
> 
> I am in favor of the first option but others should weigh in as well. Related to this, we should discuss this in the Silicon working group, e.g., tomorrow evening. Can you please liaise with the Silicon WG chairs @GregAC and @msfschaffner to make sure it gets on the agenda?

Following up on this discussion and the consensus reached in the Silicon WG 12 Dec 2023, I've created a reduced PR https://github.com/lowRISC/opentitan/pull/20675 with the SHA-2 primitive RTL only. We will hold back the changes to instantiate this in DMA and HMAC RTL so close to tapeout for later - although I've confirmed regression pass/fail rates are unaffected, coverage metrics are different obviously.

**Comment by msfschaffner** [comment]
Thanks @gdessouky - do we have an initial assessment of how big the impact on coverage metrics is?

**Comment by gdessouky** [comment]
> Thanks @gdessouky - do we have an initial assessment of how big the impact on coverage metrics is?

Total Coverage Summary 
SCORE 95.67
LINE  99.05  
COND  96.32 
TOGGLE 100.00  
**FSM   81.82**  - all metrics are reduced a bit but FSM coverage the most and below 90
BRANCH 93.13 
ASSERT 99.49
GROUP  99.86

But based on our discussions earlier, I've reorged the RTL and hope it improves coverage, besides running UNR for exclusion waivers, if we need to. Will update here on coverage after the nightly regression run concludes.

**Comment by gdessouky** [comment]
After reorganizing the RTL as suggested in #20675, coverage for HMAC did not improve significantly.  But after running UNR analysis, and updating the UNR exclusions, coverage has reached numbers on par with the current coverage.

SCORE  |  LINE   |  COND   |  TOGGLE  |   FSM    |  BRANCH  |  ASSERT  |  GROUP  |
|:-------:|:-------:|:-------:|:--------:|:--------:|:--------:|:--------:|:-------:|
| 99.62 % | 99.55 % | 98.65 % | 100.00 % | 100.00 % | 99.78 %  | 99.49 %  | 99.86 % |

Given this, I think it would be safe to update the HMAC IP with the instantiated SHA-2 primitive and clean up redundant code. WDYT @msfschaffner @vogelpi?

**Comment by gdessouky** [comment]
For DMA, pass rate is 100% with the instantiated primitive and coverage is pretty much the same.

Currently:

|  SCORE  |  LINE   |  COND   |  TOGGLE  |   FSM   |  BRANCH  |  ASSERT  |  GROUP  |
|:-------:|:-------:|:-------:|:--------:|:-------:|:--------:|:--------:|:-------:|
| 77.95 % | 96.68 % | 94.81 % | 69.65 %  | 60.00 % | 94.11 %  | 98.15 %  | 32.24 % |


With the instantiated prim:
|  SCORE  |  LINE   |  COND   |  TOGGLE  |   FSM   |  BRANCH  |  ASSERT  |  GROUP  |
|:-------:|:-------:|:-------:|:--------:|:-------:|:--------:|:--------:|:-------:|
| 78.69 % | 97.03 % | 94.99 % | 69.65 %  | 64.00 % | 94.56 %  | 98.15 %  | 32.45 % |

So again here, I'd say it's safe to update the DMA IP with the instantated SHA-2 primitive - @Razer6 @neeraj-rv @msfschaffner WDYT?

File: hw/ip/hmac/hmac.core
Comment (msfschaffner) [inline hw/ip/hmac/hmac.core:6]: I believe we should not be changing this description yet, since we haven't actually implemented multi-mode just yet.

**Review by msfschaffner — APPROVED** [review]
Thanks @gdessouky, I think with these coverage numbers it should be ok to merge this.

Before we do that though I'd like to loop in @andreaskurth and @Razer6 to see whether that is ok from their end as well.

File: hw/ip/hmac/hmac.core
Comment (gdessouky) [reply hw/ip/hmac/hmac.core]: thanks, i updated the description in my last forced push.

**Review by Razer6 — APPROVED** [review]
Thanks for making this. I went through the commits and this looks good to me.

**Comment by andreaskurth** [comment]
Thanks for this work and the extra effort for keeping DV maturity at the existing high level, @gdessouky! :+1:

**Review by cdgori — COMMENTED** [review]
I'm catching up on all hmac/sha2 changes since January, as I was temporarily not working on OT from then until now.  I'm working through things in reverse chronological order to make sure I understand what was done.

This LGTM.

**Pull request merged**
